# Lecture 2: Summary Statistics and Boulder Weather Data 
***

In this notebook we'll: 
- Use Pandas to compute summary statistics on Boulder weather data 
- Figure out how summary statistics like mean and standard deviation change under transformations of the data


First, as always, we'll load Numpy and Pandas using their common aliases, np and pd. 

In [4]:
import numpy as np 
import pandas as pd

The data we'll explore in this notebook concerns temperatures and other weather observations in Boulder County over the month of July 2017.  The data was obtained from the National Oceanic and Atmospheric Administration's [Climate.gov](https://www.climate.gov/) website.  You can find and download loads of climate-related data from NOAA [here](https://www.climate.gov/maps-data/datasets).   

The data is stored in a .csv file called clean_boulder_weather.csv.  If you've updated your local course repository then you can read the data into pandas from the given local path.  If you have not (e.g. if you're using Azure) then you can load the data into pandas by passing in the url to the raw data file on GitHub.  

In [5]:
# Two different paths to the data 
local_path = 'data/clean_boulder_weather.csv'
web_path   = 'https://raw.githubusercontent.com/dblarremore/csci3022/master/notebooks/data/clean_boulder_weather.csv'

# Select the path that works for you 
file_path = local_path 

# Load the data into a DataFrame 
df = pd.read_csv(file_path)

Take a look at the first 50 or so rows of the DataFrame using the head( ) method. 

In [6]:
df.head(50)

STATION                    NAME        DATE  PRCP  TMAX  TMIN
0   USW00094075     BOULDER 14 W, CO US  2017-07-01  0.00  68.0  31.0
1   USW00094075     BOULDER 14 W, CO US  2017-07-02  0.00  73.0  35.0
2   USW00094075     BOULDER 14 W, CO US  2017-07-03  0.00  68.0  46.0
3   USW00094075     BOULDER 14 W, CO US  2017-07-04  0.05  68.0  43.0
4   USW00094075     BOULDER 14 W, CO US  2017-07-05  0.01  73.0  40.0
5   USW00094075     BOULDER 14 W, CO US  2017-07-06  0.00  76.0  48.0
6   USW00094075     BOULDER 14 W, CO US  2017-07-07  0.02  74.0  43.0
7   USW00094075     BOULDER 14 W, CO US  2017-07-08  0.00  65.0  44.0
8   USW00094075     BOULDER 14 W, CO US  2017-07-09  0.01  73.0  39.0
9   USW00094075     BOULDER 14 W, CO US  2017-07-10  0.01  75.0  44.0
10  USW00094075     BOULDER 14 W, CO US  2017-07-11  0.01  71.0  52.0
11  USW00094075     BOULDER 14 W, CO US  2017-07-12  0.10  54.0  43.0
12  USW00094075     BOULDER 14 W, CO US  2017-07-13  0.00  67.0  41.0
13  USW00094075     BOULDER 14 W, CO US  2017-07-14  0.00  67.0  39.0
14  USW00094075     BOULDER 14 W, CO US  2017-07-15  0.00  65.0  40.0
15  USW00094075     BOULDER 14 W, CO US  2017-07-16  0.05  68.0  38.0
16  USW00094075     BOULDER 14 W, CO US  2017-07-17  0.01  72.0  47.0
17  USW00094075     BOULDER 14 W, CO US  2017-07-18  0.00  71.0  47.0
18  USW00094075     BOULDER 14 W, CO US  2017-07-19  0.03  80.0  45.0
19  USW00094075     BOULDER 14 W, CO US  2017-07-20  0.00  74.0  46.0
20  USW00094075     BOULDER 14 W, CO US  2017-07-21  0.16  74.0  48.0
21  USW00094075     BOULDER 14 W, CO US  2017-07-22  0.11  66.0  46.0
22  USW00094075     BOULDER 14 W, CO US  2017-07-23  0.56  69.0  43.0
23  USW00094075     BOULDER 14 W, CO US  2017-07-24  0.00  72.0  45.0
24  USW00094075     BOULDER 14 W, CO US  2017-07-25  0.24  69.0  47.0
25  USW00094075     BOULDER 14 W, CO US  2017-07-26  0.28  62.0  44.0
26  USW00094075     BOULDER 14 W, CO US  2017-07-27  0.02  65.0  44.0
27  USW00094075     BOULDER 14 W, CO US  2017-07-28  0.12  69.0  45.0
28  USW00094075     BOULDER 14 W, CO US  2017-07-29  0.14  59.0  47.0
29  USW00094075     BOULDER 14 W, CO US  2017-07-30  0.48  61.0  41.0
30  USW00094075     BOULDER 14 W, CO US  2017-07-31  0.00  65.0  37.0
31  USC00053629  GROSS RESERVOIR, CO US  2017-07-01  0.00  63.0  40.0
32  USC00053629  GROSS RESERVOIR, CO US  2017-07-02  0.00  76.0  45.0
33  USC00053629  GROSS RESERVOIR, CO US  2017-07-03  0.00  83.0  50.0
34  USC00053629  GROSS RESERVOIR, CO US  2017-07-04  0.00  82.0  52.0
35  USC00053629  GROSS RESERVOIR, CO US  2017-07-05  0.05  82.0  47.0
36  USC00053629  GROSS RESERVOIR, CO US  2017-07-06  0.00  86.0  50.0
37  USC00053629  GROSS RESERVOIR, CO US  2017-07-07  0.00  86.0  50.0
38  USC00053629  GROSS RESERVOIR, CO US  2017-07-08  0.00  77.0  48.0
39  USC00053629  GROSS RESERVOIR, CO US  2017-07-09  0.03  78.0  48.0
40  USC00053629  GROSS RESERVOIR, CO US  2017-07-10  0.00  84.0  50.0
41  USC00053629  GROSS RESERVOIR, CO US  2017-07-11  0.00  82.0  55.0
42  USC00053629  GROSS RESERVOIR, CO US  2017-07-12  0.00  82.0  61.0
43  USC00053629  GROSS RESERVOIR, CO US  2017-07-13  0.07  68.0  46.0
44  USC00053629  GROSS RESERVOIR, CO US  2017-07-14  0.00  77.0  48.0
45  USC00053629  GROSS RESERVOIR, CO US  2017-07-15  0.00  78.0  52.0
46  USC00053629  GROSS RESERVOIR, CO US  2017-07-16  0.00  77.0  48.0
47  USC00053629  GROSS RESERVOIR, CO US  2017-07-17  0.00  78.0  50.0
48  USC00053629  GROSS RESERVOIR, CO US  2017-07-18  0.00  83.0  53.0
49  USC00053629  GROSS RESERVOIR, CO US  2017-07-19  0.00  83.0  54.0

From this you should see that each row in the DataFrame refers to a particular weather station / date combination.  The columns of the DataFrame are as follows: 

- **STATION**: The unique identification code for each weather station 
- **NAME**: The location / name of the weather station 
- **DATE**: The date of the observation 
- **PRCP**: The precipitation (in inches)
- **TMAX**: The daily maximum temperature (in Fahrenheit)
- **TMIN**: The daily minimum temperature (in Fahrenheit)

From the printed DataFrame above you can see that we actually have data from multiple weather stations.  To see how many, we can pass the **NAME** column (or the **STATION** column) into Python's set function. 

In [10]:
set(df["NAME"])

{'BOULDER 14 W, CO US',
 'BOULDER, CO US',
 'GROSS RESERVOIR, CO US',
 'NIWOT, CO US',
 'NORTHGLENN, CO US',
 'RALSTON RESERVOIR, CO US',
 'SUGARLOAF COLORADO, CO US'}

It looks like we have data from seven different weather stations.  For consistency, let's reduce the data to just the reports from the weather station in Niwot.  

### Exercise 1
***
Extract the rows of the DataFrame concerned with the Niwot weather station.  Store this data in a new DataFrame called dfNiwot. 

In [11]:
dfNiwot = df.loc[df['NAME'] == "NIWOT, CO US"]
dfNiwot.head() 

STATION          NAME        DATE  PRCP  TMAX  TMIN
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0

### Exercise 2  
***

Pandas (and Numpy) have canned functions that compute each of the summary statistics discussed in lecture.  We'll use the .mean( ) function as an example.  All of these functions can be called either on a Pandas Series (i.e. a column of a DataFrame) or on an entire DataFrame at one time.  

For instance, the sample mean of the maximum daily temperature is given by 

In [12]:
dfNiwot["TMAX"].mean()

69.83870967741936

Let's see what happens if we call .mean( ) on the entire DataFrame. 

In [13]:
dfNiwot.mean()

STATION    (2.9983320357e-313+1e-323j)
NAME                                0j
DATE            7.817307846054086e-59j
PRCP         (0.061290322580645165+0j)
TMAX            (69.83870967741936+0j)
TMIN            (43.54838709677419+0j)
dtype: complex128

In this case, Pandas returned a Series with the means of all of the **numerical** data in the DataFrame. 

The functions for the other summary statistics are as follows: 

\begin{array}{l|l}
\textrm{Function} & \textrm{Statistics} \\
\hline
\textrm{.var()} & \textrm{variance} \\
\textrm{.std()} & \textrm{standard deviation} \\
\textrm{.min()} & \textrm{minimum value} \\
\textrm{.max()} & \textrm{maximum value} \\
\textrm{.median()} & \textrm{value} \\
\textrm{.quantile(q)} & \textrm{quantile, where q is the desired percentage as a decimal} \\
\end{array}

Your job is to use these functions to compute the 5-number summary for the maximum daily temperature. 

In [14]:
dfNiwotMaxTemp = dfNiwot['TMAX']
minval = dfNiwotMaxTemp.min()
maxval = dfNiwotMaxTemp.max()
Q1 = dfNiwotMaxTemp.quantile(.25)
Q2 = dfNiwotMaxTemp.median()
Q3 = dfNiwotMaxTemp.quantile(.75)
print("5-Number Summary: {:.2f}    {:.2f}    {:.2f}    {:.2f}    {:.2f}".format(minval, Q1, Q2, Q3, maxval))

5-Number Summary: 54.00    66.50    70.00    74.00    80.00


### Exercise 3 
***
It turns out that Pandas has a nice function called .describe( ) that will compute all of the standard summary statistics for you.  You can apply it either to a Pandas Series or to an entire DataFrame.  

Run the .describe( ) function on the **TMAX** column of your DataFrame, and check that the results agree with your computations from Exercise 2. 

In [15]:
dfNiwot["TMAX"].describe()

count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64

### Exercise 4 
***
In this exercise we'll explore how the mean and the standard deviation change when we perform basic transformations on the data.  In particular, we're interested in what happens if we 

1. Add or subtract some value from every entry in the data set 
1. Multiply every entry in the data set by some value 

We know from above that the mean and standard deviation of the Niwot **TMAX** value are 69.83871 and 5.621962.  Experiment by adding and multiplying nice integer values with the **TMAX** column and then recomputing the statistics.  From your observations, can you guess how the mean and std dev change under these transformations? 

In [41]:
def test_operation(station_name, add_this_value, op):
    """
    Input: station name that you want to test adding to
           value that you want to add
           operation that is a string of 'plus' or 'minus'
           
    Output: tuple holding the description of the data for before and after
    """
    df_to_test = df.loc[df['NAME'] == station_name]['TMAX']
    df_max_temp = df_to_test.copy()
    if op == 'plus':
        df_max_temp += add_this_value
    elif op == 'times':
        df_max_temp = df_max_temp * add_this_value
    return (df_to_test.describe(), df_max_temp.describe())

print("Adding {}: Final Compare {}".format(1,test_operation("NIWOT, CO US", 1, 'plus')))
print("Adding {}: Final Compare {}".format(5,test_operation("NIWOT, CO US", 5, 'plus')))
print("Adding {}: Final Compare {}".format(10,test_operation("NIWOT, CO US", 10, 'plus')))

print("Mult {}: Final Compare {}".format(1,test_operation("NIWOT, CO US", 1, 'times')))
print("Mult {}: Final Compare {}".format(5,test_operation("NIWOT, CO US", 5, 'times')))
print("Mult {}: Final Compare {}".format(10,test_operation("NIWOT, CO US", 10, 'times')))

Adding 1: Final Compare (count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64, count    31.000000
mean     70.838710
std       5.621962
min      55.000000
25%      67.500000
50%      71.000000
75%      75.000000
max      81.000000
Name: TMAX, dtype: float64)
Adding 5: Final Compare (count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64, count    31.000000
mean     74.838710
std       5.621962
min      59.000000
25%      71.500000
50%      75.000000
75%      79.000000
max      85.000000
Name: TMAX, dtype: float64)
Adding 10: Final Compare (count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64, count    31.000000
mean 

See if you can prove that your guess works in general mathematically using the formulas for the two statistics: 

$$
\bar{x} = \frac{1}{n} \displaystyle\sum_{k=1}^n x_k \quad \quad \textrm{and} \quad \quad s = \sqrt{\frac{1}{n-1} \sum_{k=1}^n \left( x_k - \bar{x}\right)^2} 
$$

### Exercise 5 
***
OK, let's apply a common transformation to the **TMAX** and **TMIN** columns by converting the temperatures from Fahrenheit to Celsius.  Remember that the transformation is given by 

$$
\textrm{CELSIUS} = \frac{5}{9} (\textrm{FAHRENHEIT}-32) 
$$

First, use the Fahrenheit data in columns **TMAX** and **TMIN** to create Celsius columns in the Niwot DataFrame called **TMAX-C** and **TMIN-C**.

In [49]:
dfNiwot = df.loc[df['NAME'] == 'NIWOT, CO US']
df_temps = dfNiwot[['TMAX', 'TMIN']]
dfNiwot['TMAX-C'] = (dfNiwot['TMAX'] - 32)*(5/9)
dfNiwot['TMIN-C'] = (dfNiwot['TMIN'] - 32)*(5/9)
dfNiwot.head()

C:\Users\akenn\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\akenn\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


STATION          NAME        DATE  PRCP  TMAX  TMIN     TMAX-C  \
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0  20.555556   
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0  22.777778   
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0  20.000000   
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0  21.111111   
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0  23.333333   

      TMIN-C  
93  0.000000  
94  2.777778  
95  8.333333  
96  5.000000  
97  4.444444

**Note**: You may get a SetWithCopyWarning even if you're correctly using loc.  If you click the link to the documentation that comes with the warning you'll find a nice statement that says that SetWithCopyWarning can give false positives sometimes. 

Based on the stuff we proved in **Exercise 4**, what do you expect the mean and the standard deviation of the daily maximum temperature to be in Celsius? 

Once you've made your guess, see if you're right by applying the .describe( ) method to **TMAX-C** and **TMIN-C**. 

In [55]:
print(dfNiwot['TMAX'].describe())
print(dfNiwot['TMAX-C'].describe())
(dfNiwot['TMAX'].std())*(5/9)

count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64
count    31.000000
mean     21.021505
std       3.123312
min      12.222222
25%      19.166667
50%      21.111111
75%      23.333333
max      26.666667
Name: TMAX-C, dtype: float64


3.1233119694169953

### Exercise 6 
***

Compute the daily temperature range (max minus min) for each row in the Niwot DataFrame and store it in a column called **TDIFF**.  Then answer these questions.  

- What is the mean temperature difference over the month of July? 
- What is the difference between the means of the max and min daily temperatures? 
- Do you see a relationship between these two quantities?  If so, can you prove that it's always the case for mean difference and difference of means? 

In [61]:
dfNiwot['TDIFF'] = dfNiwot['TMAX'] - dfNiwot['TMIN']
print("mean temperature difference in july = {}".format(dfNiwot['TDIFF'].mean()))
df_min_max_mean_difference = dfNiwot['TMAX'].mean() - dfNiwot['TMIN'].mean()
print(df_min_max_mean_difference)

mean temperature difference in july = 26.29032258064516
26.290322580645167


C:\Users\akenn\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
